src/coreclr/src/vm/weakreferencenative.cpp

```diff
@@ -132,21 +133,35 @@ IWeakReference* GetWinRTWeakReference(OBJECTREF* pObject)
 
     MethodTable* pMT = (*pObject)->GetMethodTable();
 
-    // If the object is not an RCW, then we do not want to use a WinRT weak reference to it
-    if (!pMT->IsComObjectType())
-    {
-        return nullptr;
-    }
+    SafeComHolder<IWeakReferenceSource> pWeakReferenceSource(nullptr);
 
-    // If the object is a managed type deriving from a COM type, then we also do not want to use a WinRT
+    // If the object is not an RCW, then we do not want to use a native COM weak reference to it
+    // If the object is a managed type deriving from a COM type, then we also do not want to use a native COM
     // weak reference to it.  (Otherwise, we'll wind up resolving IWeakReference-s back into the CLR
     // when we don't want to have reentrancy).
-    if (pMT != g_pBaseCOMObject && pMT->IsExtensibleRCW())
+    if (pMT->IsComObjectType()
+     && (pMT == g_pBaseCOMObject || !pMT->IsExtensibleRCW()))
     {
-        return nullptr;
+        pWeakReferenceSource = reinterpret_cast<IWeakReferenceSource*>(GetComIPFromObjectRef(pObject, IID_IWeakReferenceSource, false /* throwIfNoComIP */));
+    }
+#ifdef FEATURE_COMWRAPPERS
+    else
+    {
+
+        IUnknown* unknown = ComWrappersNative::GetIdentityForObject(pObject);
```

Comment:
Let's enrich the `ComWrappersNative::GetIdentityForObject()` API to be symmetrical with `GetComIPFromObjectRef()`. Do this would also allow us to take perform the QI immediately instead of returning null and doing yet another check here. The API still returns `null` to indicate no.

```C++
IUnknown* ComWrappersNative::GetInterfaceFromObject(_In_ OBJECTREF* objectPROTECTED, _In_ REFIID riid);
```